### PR TITLE
[IMP] mrp: improvements in Partial production with Quality Check

### PR DESCRIPTION
**Pull request by pgu-odoo**
In this commit -
If final product is not serial tracking, then system will suggest the units to produce in next WO according
to the previous one.
For example - If we have MO like, MO0001 (5 qty)-> WO01, WO02, WO03

Then in WO01, if only 2 qtys are produced then, in WO02, 2 qty will be sugested insteasd of 5.

Task-2088002

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/tests/test_workorder_operation.py [hunk 1]
```diff
@@ -1297,6 +1313,51 @@ def setUpClass(cls):
...
+        cls.operation_grinding = cls.env['mrp.routing.workcenter'].create({
+            'sequence': 1,
+            'name': 'Grinding',
+            'workcenter_id': cls.workcenter_grinder.id,
+            'routing_id': cls.routing_cake_making.id,
+            'batch': 'yes'
+        })
+        cls.operation_mixing = cls.env['mrp.routing.workcenter'].create({
+            'sequence': 1,
+            'name': 'Grinding',
```
Comment (amoyaux) [inline addons/mrp/tests/test_workorder_operation.py]: Mixing

File: addons/mrp/tests/test_workorder_operation.py [hunk 2]
```diff
@@ -1297,6 +1313,51 @@ def setUpClass(cls):
...
+            'workcenter_id': cls.workcenter_mixer.id,
+            'routing_id': cls.routing_cake_making.id,
+            'batch': 'yes'
+        })
+        cls.operation_baking = cls.env['mrp.routing.workcenter'].create({
+            'sequence': 1,
+            'name': 'Grinding',
```
Comment (amoyaux) [inline addons/mrp/tests/test_workorder_operation.py]: Baking

File: addons/mrp/tests/test_workorder_operation.py [hunk 3]
```diff
@@ -1484,3 +1545,109 @@ def test_merge_lot(self):
...
+        self.fruit_cake.tracking = 'lot'
+        lot_1 = self.env['stock.production.lot'].create({
+            'product_id': self.fruit_cake.id,
+            'company_id': self.env.company.id,
+        })
+        lot_2 = self.env['stock.production.lot'].create({
+            'product_id': self.fruit_cake.id,
+            'company_id': self.env.company.id,
+        })
+        produced_msg = "wrong qty is produced."
```
Comment (amoyaux) [inline addons/mrp/tests/test_workorder_operation.py]: "Wrong quantity produced."

File: addons/mrp/tests/test_workorder_operation.py [hunk 4]
```diff
@@ -1484,3 +1545,109 @@ def test_merge_lot(self):
...
+        processing_msg = "wrong qty is suggested to produce."
```
Comment (amoyaux) [inline addons/mrp/tests/test_workorder_operation.py]: "Wrong to produce quantity suggested."

File: addons/mrp/wizard/change_production_qty.py [hunk 5]
```diff
@@ -92,6 +92,9 @@ def change_prod_qty(self):
                                  operation.workcenter_id.time_stop +
                                  cycle_number * operation.time_cycle * 100.0 / operation.workcenter_id.time_efficiency)
                 quantity = wo.qty_production - wo.qty_produced
+                is_not_first_order = bool(production.workorder_ids.filtered(lambda l: l.next_work_order_id == wo))
```
Comment (amoyaux) [inline addons/mrp/wizard/change_production_qty.py]: wo in production.workorder_ids.mapped('next_workorder_id')

File: addons/mrp/wizard/change_production_qty.py [hunk 6]
```diff
@@ -92,6 +92,9 @@ def change_prod_qty(self):
...
+                if wo.operation_id.batch and wo.state in ['processing', 'ready'] and is_not_first_order:
+                    quantity = wo.qty_producing
```
Comment (amoyaux) [inline addons/mrp/wizard/change_production_qty.py]: You could probably call _defaults_from_finished_workorder_line. Anyway this would be erase 3 lines after since you didn't use elif. Check if we can get rid of this.

File: addons/mrp/wizard/mrp_product_produce.py [hunk 7]
```diff
@@ -55,6 +55,7 @@ def default_get(self, fields):
         'finished_product_produce_id', string='By-products')
     production_id = fields.Many2one('mrp.production', 'Manufacturing Order',
         required=True, ondelete='cascade')
+    qty_remaining = fields.Float('Quantity To Be Produced', compute='_get_todo', digits='Product Unit of Measure')
```
Comment (amoyaux) [inline addons/mrp/wizard/mrp_product_produce.py]: use the _compute_pending_production or modify _get_todo to _compute_get_todo and use this field instead of call to function

File: addons/mrp/models/mrp_workorder.py [hunk 8]
```diff
@@ -421,18 +431,40 @@ def record_production(self):
         # Update workorder quantity produced
         self.qty_produced += self.qty_producing
 
+        rounding = self.production_id.product_uom_id.rounding
+        previous_wo = self.env['mrp.workorder'].search([
+            ('next_work_order_id', '=', self.id)])
         # Suggest a finished lot on the next workorder
-        if self.next_work_order_id and self.product_tracking != 'none' and (not self.next_work_order_id.finished_lot_id or self.next_work_order_id.finished_lot_id == self.finished_lot_id):
-            self.next_work_order_id._defaults_from_finished_workorder_line(self.finished_workorder_line_ids)
+        if self.next_work_order_id and self.product_tracking != 'none' and (not self.next_work_order_id.finished_lot_id or self.next_work_order_id.finished_lot_id == self.finished_lot_id) and self.next_work_order_id.state != 'done':
```
Comment (amoyaux) [inline addons/mrp/models/mrp_workorder.py]: We could probably remove the if now that quantity are taken into account. And do all the job in _defaults_from_finished_workorder_line

**Review by amoyaux — CHANGES_REQUESTED** [review]

File: addons/mrp/models/mrp_workorder.py [hunk 9]
```diff
@@ -421,18 +431,40 @@ def record_production(self):
         # Update workorder quantity produced
         self.qty_produced += self.qty_producing
 
+        rounding = self.production_id.product_uom_id.rounding
```
[the remark this answers is not in the document]
Comment (pgu-odoo) [reply addons/mrp/models/mrp_workorder.py]: rounding is also used to check if production is done or not at https://github.com/odoo/odoo/blob/13.0/addons/mrp/models/mrp_workorder.py#L436
removing the diff for this line.

File: addons/mrp/models/mrp_abstract_workorder.py [hunk 10]
```diff
@@ -90,6 +90,8 @@ def _update_workorder_lines(self):
                 self.production_id.product_uom_id,
                 round=False
             )
+            if float_compare(new_qty, 0.0, precision_rounding=rounding) <= 0:
```
[the remark this answers is not in the document]
Comment (pgu-odoo) [reply addons/mrp/models/mrp_abstract_workorder.py]: We are suggesting the quantity to 0.0 if previously produced quantity is already produced, but still production is not completed.
Issue - 
1) Component remaining quantity is not updating on on change of qty_producing if once we have suggested 0.0 (and if work line is not created). issue mentioned on pad -
https://drive.google.com/file/d/1aKeeSuMYkc3v8PIf6Bij5YYvXpYpwdEz/view
Component remaining quantity is updated as -
https://github.com/odoo/enterprise/blob/13.0/mrp_workorder/models/mrp_workorder.py#L101

File: addons/mrp/models/mrp_workorder.py (hunk 8 above)
Comment (pgu-odoo) [reply addons/mrp/models/mrp_workorder.py]: method _defaults_from_finished_workorder_line is responsible for assigning the lot id if exist in current wo and returns the True or False if candidate is found or not.
And this code is -
1) To suggest the previously produced quantity even if lot id is not present.
So it this case we can not say that candidate is found but need to suggest the quantity thats why change the code here.

File: addons/mrp/models/mrp_workorder.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (amoyaux) [reply addons/mrp/models/mrp_workorder.py]: I was not speaking about just this line sorry, I mean all the code block should move to _defaults_from_finished_workorder_line.
I know that it currently only assign a lot or SN to the workorder but I think it should also become responsible for the quantity

File: addons/mrp/models/mrp_workorder.py [hunk 11]
```diff
@@ -421,16 +440,31 @@ def record_production(self):
...
+            if self.next_work_order_id and self.next_work_order_id.state != 'done' and self.product_tracking == 'none':
+                self.next_work_order_id.qty_producing = self.qty_produced - self.next_work_order_id.qty_produced
+                self.next_work_order_id._apply_update_workorder_lines()
+            # To update the qty_producing of current WO (Used in partial production)
+            # To make the qty_producing = 0 if product tracking is LOT and previously produced qty is already processed in current WO.
+            suggested_qty = self.qty_remaining
+            if previous_wo:
+                suggested_qty = previous_wo.qty_produced - self.qty_produced
+            self.qty_producing = 0.0 if suggested_qty < 0 else suggested_qty
+
```
Comment (amoyaux) [inline addons/mrp/models/mrp_workorder.py]: you should also move this part

Comment (amoyaux) [inline addons/mrp/models/mrp_workorder.py]: you should also move this part

File: addons/mrp/wizard/mrp_product_produce.py [hunk 12]
```diff
@@ -55,6 +55,7 @@ def default_get(self, fields):
         'finished_product_produce_id', string='By-products')
     production_id = fields.Many2one('mrp.production', 'Manufacturing Order',
         required=True, ondelete='cascade')
+    qty_remaining = fields.Float('Quantity To Be Produced', compute='_compute_pending_production', digits='Product Unit of Measure')
```
Comment (Whenrow) [inline addons/mrp/wizard/mrp_product_produce.py]: This fields is not set in the compute function and not use either

File: addons/mrp/models/mrp_workorder.py [hunk 13]
```diff
@@ -169,12 +169,19 @@ def _onchange_finished_lot_id(self):
         """When the user changes the lot being currently produced, suggest
         a quantity to produce consistent with the previous workorders. """
         previous_wo = self.env['mrp.workorder'].search([
-            ('next_work_order_id', '=', self.id)
+            ('next_work_order_id', '=', self.ids[0])
```
Comment (Whenrow) [inline addons/mrp/models/mrp_workorder.py]: Why this change ?

File: addons/mrp/models/mrp_workorder.py [hunk 14]
```diff
@@ -169,12 +169,19 @@ def _onchange_finished_lot_id(self):
...
         ])
         if previous_wo:
+            suggested_qty = 0.0
             line = previous_wo.finished_workorder_line_ids.filtered(lambda line: line.product_id == self.product_id and line.lot_id == self.finished_lot_id)
             if line:
-                self.qty_producing = line.qty_done
+                suggested_qty = line.qty_done
+                finished_line = self.finished_workorder_line_ids.filtered(lambda line: line.product_id == self.product_id and line.lot_id == self.finished_lot_id)
+                if finished_line:
+                    suggested_qty = line.qty_done - finished_line.qty_done
```
Comment (Whenrow) [inline addons/mrp/models/mrp_workorder.py]: as float field on empty recordset will be equal to 0, you can shorten this with 
```python   
finished_line = self.finished_workorder_line_ids.filtered(lambda line: line.product_id == self.product_id and line.lot_id == self.finished_lot_id) 
suggested_qty = line.qty_done - finished_line.qty_done
```

File: addons/mrp/models/mrp_workorder.py [hunk 15]
```diff
@@ -169,12 +169,19 @@ def _onchange_finished_lot_id(self):
...
+
+            suggested_qty = min(suggested_qty, self.qty_remaining)
+            self.qty_producing = 0.0 if suggested_qty < 0 else suggested_qty
```
Comment (Whenrow) [inline addons/mrp/models/mrp_workorder.py]: self.qty_producing = max(0.0, suggested_qty)

File: addons/mrp/models/mrp_workorder.py [hunk 16]
```diff
@@ -377,26 +384,32 @@ def _refresh_wo_lines(self):
...
+        suggested_lot = False
+        suggested_qty = finished_wo.qty_produced - self.qty_produced
+
+        if self.product_tracking != 'none':
+            for r_line in finished_wo.finished_workorder_line_ids:
+                # see which lot we could suggest and its related qty_producing
+                if not r_line.lot_id:
+                    if self.product_tracking == 'serial':
+                        suggested_qty = 1
+                    continue
```
Comment (Whenrow) [inline addons/mrp/models/mrp_workorder.py]: I think It's useless. 
At creation, the quantity is set to 1. Once the production start, all workorder lines have 1 as quantity done

File: addons/mrp/models/mrp_workorder.py (hunk 13 above)
Comment (pgu-odoo) [reply addons/mrp/models/mrp_workorder.py]: At onchange self is like mrp.workorder(<NewId origin=4>,) here. And is not working.

Before this, correct qty_producing was not suggested according to selected lot.
In this -
For example if we have MO with 2 WO, then if production is done as-
WO1 - L001 -> 2 Qty produced
            L002 -> 3 Qty produced
Then, in WO2 if user selects L001 then 2 should be suggested qty_producing and 3 if L002 is selected.

**Comment by amoyaux** [comment]
@Whenrow Could you check my fixup

**Comment by lathama** [comment]
@pgu-odoo  knowing that 13 is no longer updated, is this PR still needed?

**Comment by C3POdoo** [comment]
Dear @pgu-odoo,

Thank you for your contribution but the version 13.0 is no longer supported.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense for the upper version, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._
